niff-1.02/niffsdk1.02/riffio/doc/man/rfile.man
.rn '' }`
''' $RCSfile$$Revision$$Date$
'''
''' $Log$
'''
.de Sh
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp
.if t .sp .5v
.if n .sp
..
.de Ip
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.de Vb
.ft CW
.nf
.ne \\$1
..
.de Ve
.ft R

.fi
..
'''
'''
'''     Set up \*(-- to give an unbreakable dash;
'''     string Tr holds user defined translation string.
'''     Bell System Logo is used as a dummy character.
'''
.tr \(*W-|\(bv\*(Tr
.ie n \{\
.ds -- \(*W-
.ds PI pi
.if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\" diablo 12 pitch
.ds L" ""
.ds R" ""
.ds L' '
.ds R' '
'br\}
.el\{\
.ds -- \(em\|
.tr \*(Tr
.ds L" ``
.ds R" ''
.ds L' `
.ds R' '
.ds PI \(*p
'br\}
.\"	If the F register is turned on, we'll generate
.\"	index entries out stderr for the following things:
.\"		TH	Title 
.\"		SH	Header
.\"		Sh	Subsection 
.\"		Ip	Item
.\"		X<>	Xref  (embedded
.\"	Of course, you have to process the output yourself
.\"	in some meaninful fashion.
.if \nF \{
.de IX
.tm Index:\\$1\t\\n%\t"\\$2"
..
.nr % 0
.rr F
.\}
.TH rfile 3 "riffio" "14/Jun/96" "RIFFIO Documentation"
.IX Title "rfile 3"
.UC
.IX Name "rfile - RIFFIOFile routines"
.if n .hy 0
.if n .na
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.de CQ          \" put $1 in typewriter font
.ft CW
'if n "\c
'if t \\&\\$1\c
'if n \\&\\$1\c
'if n \&"
\\&\\$2 \\$3 \\$4 \\$5 \\$6 \\$7
'.ft R
..
.\" @(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2
.	\" AM - accent mark definitions
.bd B 3
.	\" fudge factors for nroff and troff
.if n \{\
.	ds #H 0
.	ds #V .8m
.	ds #F .3m
.	ds #[ \f1
.	ds #] \fP
.\}
.if t \{\
.	ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.	ds #V .6m
.	ds #F 0
.	ds #[ \&
.	ds #] \&
.\}
.	\" simple accents for nroff and troff
.if n \{\
.	ds ' \&
.	ds ` \&
.	ds ^ \&
.	ds , \&
.	ds ~ ~
.	ds ? ?
.	ds ! !
.	ds /
.	ds q
.\}
.if t \{\
.	ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.	ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.	ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.	ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.	ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.	ds ? \s-2c\h'-\w'c'u*7/10'\u\h'\*(#H'\zi\d\s+2\h'\w'c'u*8/10'
.	ds ! \s-2\(or\s+2\h'-\w'\(or'u'\v'-.8m'.\v'.8m'
.	ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.	ds q o\h'-\w'o'u*8/10'\s-4\v'.4m'\z\(*i\v'-.4m'\s+4\h'\w'o'u*8/10'
.\}
.	\" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds v \\k:\h'-(\\n(.wu*9/10-\*(#H)'\v'-\*(#V'\*(#[\s-4v\s0\v'\*(#V'\h'|\\n:u'\*(#]
.ds _ \\k:\h'-(\\n(.wu*9/10-\*(#H+(\*(#F*2/3))'\v'-.4m'\z\(hy\v'.4m'\h'|\\n:u'
.ds . \\k:\h'-(\\n(.wu*8/10)'\v'\*(#V*4/10'\z.\v'-\*(#V*4/10'\h'|\\n:u'
.ds 3 \*(#[\v'.2m'\s-2\&3\s0\v'-.2m'\*(#]
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.ds oe o\h'-(\w'o'u*4/10)'e
.ds Oe O\h'-(\w'O'u*4/10)'E
.	\" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.	\" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.	ds : e
.	ds 8 ss
.	ds v \h'-1'\o'\(aa\(ga'
.	ds _ \h'-1'^
.	ds . \h'-1'.
.	ds 3 3
.	ds o a
.	ds d- d\h'-1'\(ga
.	ds D- D\h'-1'\(hy
.	ds th \o'bp'
.	ds Th \o'LP'
.	ds ae ae
.	ds Ae AE
.	ds oe oe
.	ds Oe OE
.\}
.rm #[ #] #H #V #F C
.SH "NAME"
.IX Header "NAME"
rfile \- RIFFIOFile routines
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Ip "\(bu" 4
.IX Item "\(bu"
\fIRIFFIOFileNew()\fR
.Ip "\(bu" 4
.IX Item "\(bu"
\fIRIFFIOFileInit()\fR
.Ip "\(bu" 4
.IX Item "\(bu"
\fIRIFFIOFileDelete()\fR
.Ip "\(bu" 4
.IX Item "\(bu"
\fIRIFFIOFileGetFormType()\fR
.Ip "\(bu" 4
.IX Item "\(bu"
\fIRIFFIORead()\fR
.Ip "\(bu" 4
.IX Item "\(bu"
\fIRIFFIOWrite()\fR
.Ip "\(bu" 4
.IX Item "\(bu"
\fIRIFFIOSeek()\fR
.Ip "\(bu" 4
.IX Item "\(bu"
\fIRIFFIOTell()\fR
.SH "RIFFIOFileNew"
.IX Header "RIFFIOFileNew"
Return a pointer to a new RIFFIOFile,
or null on failure
.PP
.Vb 2
\&     RIFFIOFile *
\&     RIFFIOFileNew(void)
.Ve
.Sh "\s-1OBLIGATIONS\s0"
.IX Subsection "\s-1OBLIGATIONS\s0"
The file cannot be operated on until \fIRIFFIOFileInit()\fR is called.
.SH "RIFFIOFileInit"
.IX Header "RIFFIOFileInit"
Provide a NIFFIOFile with a user-defined I/O callbacks.
.PP
.Vb 7
\&     RIFFIOSuccess
\&     RIFFIOFileInit(RIFFIOFile *prf,
\&                    void *pvUserFile,
\&                    RIFFIOReader reader,
\&                    RIFFIOWriter writer,
\&                    RIFFIOSeeker seeker,
\&                    RIFFIOTeller teller)
.Ve
\fIRIFFFIOFileInit()\fR should be called immediately after \fINIFFIOFileNew()\fR
.Sh "\s-1ENTRY\s0"
.IX Subsection "\s-1ENTRY\s0"
.Ip "*<prf>" 4
.IX Item "*<prf>"
is the RIFFIOFile that is being initialized.
.Ip "*<pvUserFile>" 4
.IX Item "*<pvUserFile>"
points to a user representation of a file that will be passed to
user-defined reader, writer, seeker, and teller functions.
.Ip "*<reader>,<writer>,<seeker>,<teller>" 4
.IX Item "*<reader>,<writer>,<seeker>,<teller>"
are user-defined functions that perform operations on <*pvUserFile>.
.SH "RIFFIOFileDelete"
.IX Header "RIFFIOFileDelete"
Free the memory allocated to a RIFFIOFile.
.PP
.Vb 2
\&     void
\&     RIFFIOFileDelete(RIFFIOFile *prf)
.Ve
.SH "RIFFIOFileGetFormType"
.IX Header "RIFFIOFileGetFormType"
Return the form type (UNKNOWN, RIFF, or RIFX)
of a RIFFIOFile
.PP
.Vb 2
\&     RIFFIOFormType 
\&     RIFFIOFileGetFormType(RIFFIOFile *prf)
.Ve
The form type is unknown until the first form chunk of a RIFFIOFile
is descended into or created.
.SH "RIFFIORead"
.IX Header "RIFFIORead"
Read bytes from a RIFF file.
.PP
.Vb 2
\&     long 
\&     RIFFIORead(RIFFIOFile *prf, void *pvBuffer, long nBytes)
.Ve
.Sh "\s-1ENTRY\s0"
.IX Subsection "\s-1ENTRY\s0"
T <*pvBuffer> must be allocated to hold <nBytes>.
.Sh "\s-1EXIT\s0"
.IX Subsection "\s-1EXIT\s0"
.Ip "\(bu" 4
.IX Item "\(bu"
Up to <nBytes> have been read from <*prf> into <*pvBuffer>.
.Sh "\s-1RETURN\s0"
.IX Subsection "\s-1RETURN\s0"
.Ip "\(bu" 4
.IX Item "\(bu"
The number of bytes actually read,
.Ip "\(bu" 4
.IX Item "\(bu"
0 on end of file,
.Ip "\(bu" 4
.IX Item "\(bu"
-1 on error.
.SH "RIFFIOWrite"
.IX Header "RIFFIOWrite"
Write bytes to a RIFF file.
.PP
.Vb 2
\&     long 
\&     RIFFIOWrite(RIFFIOFile *prf, void *pvBuffer, long nBytes)
.Ve
Write <nBytes> from <*pvBuffer> to <*prf>.
.Sh "\s-1RETURN\s0"
.IX Subsection "\s-1RETURN\s0"
The number of bytes actually written.
.SH "RIFFIOSeek"
.IX Header "RIFFIOSeek"
Seek to a location in a RIFF file.
.PP
.Vb 2
\&     RIFFIOSuccess
\&     RIFFIOSeek(RIFFIOFile *prf, RIFFIOOffset offset, RIFFIOSeekOrigin origin)
.Ve
Seek <offset> bytes relative to <origin>.
.PP
<origin> may be one of
.PP
RIFFIO_SEEK_SET:
beginning of file
.PP
RIFFIO_SEEK_CUR:
current file position
.PP
RIFFIO_SEEK_END:
end of file
.Sh "\s-1ERRORS\s0"
.IX Subsection "\s-1ERRORS\s0"
On failure, the file position is undefined.
.SH "RIFFIOTell"
.IX Header "RIFFIOTell"
Return the current RIFF file position.
.PP
.Vb 2
\&     long
\&     RIFFIOTell(RIFFIOFile *prf)
.Ve
A file's position is measured in bytes from the
beginning of the file.

.rn }` ''


niff-1.02/niffsdk1.02/riffio/src/lib/riffio/rfile.c
#ifndef lint
static char rcsid[] =
 "$Id: rfile.c,v 1.2 1996/05/31 06:02:16 tim Exp $";
#endif
/*
 * Public Domain 1995,1996 Timothy Butler
 *
 * THIS DOCUMENT IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
 *
 */
/***************************************************************************/
/*
 * NAME
 * ====
 * rfile - RIFFIOFile routines
 *
 * SYNOPSIS
 * ======== 
 *
 * - RIFFIOFileNew()
 * - RIFFIOFileInit()
 * - RIFFIOFileDelete()
 *
 * - RIFFIOFileGetFormType()
 *
 * - RIFFIORead()
 * - RIFFIOWrite()
 * - RIFFIOSeek()
 * - RIFFIOTell()
 *
 */
/***************************************************************************/

#include <assert.h>
#include <stdlib.h>

#include "riffio.h"

/*
 * Create some bogus callbacks to catch anyone who
 * forgets to initialize a RIFFIOFile
 */
#ifndef NDEBUG

static
long 
_bogus_reader(void *vpUserFile, void *vpBuffer, long nBytes)
{
    assert(0);
}

static
long 
_bogus_writer(void *vpUserFile, void *vpBuffer, long nBytes)
{
    assert(0);
}

static 
RIFFIOSuccess 
_bogus_seeker(void *vpUserFile, 
              RIFFIOOffset offset, 
              RIFFIOSeekOrigin origin)
{
    assert(0);
}

static
long
_bogus_teller(void *vpUserFile)
{
    assert(0);
}


#endif

/***************************************************************************/
/*
 * RIFFIOFileNew
 * =============
 * Return a pointer to a new RIFFIOFile,
 * or null on failure
 */
RIFFIOFile *
RIFFIOFileNew(void)
/*
 * OBLIGATIONS
 * -----------
 * The file cannot be operated on until RIFFIOFileInit() is called.
 */
/***************************************************************************/
{
    char strModule[]= "RIFFIOFileNew";
    RIFFIOFile *prfNew;
    
    prfNew = (RIFFIOFile *) malloc(sizeof(RIFFIOFile));
    if (prfNew == 0)
    {
        RIFFIOError(strModule, "Failed to allocate memory");
        return (RIFFIOFile *) 0;
    }
    
    /*
     * Set some traps for a user who forgets to call RIFFIOFileInit()
     */
#ifndef NDEBUG
    prfNew->read  = _bogus_reader;
    prfNew->write = _bogus_writer;
    prfNew->seek  = _bogus_seeker;
    prfNew->tell  = _bogus_teller;
#endif

    return prfNew;
}
/***************************************************************************/
/*
 * RIFFIOFileInit
 * ==============
 * Provide a NIFFIOFile with a user-defined I/O callbacks.
 */
RIFFIOSuccess
RIFFIOFileInit(RIFFIOFile *prf,
               void *pvUserFile,
               RIFFIOReader reader,
               RIFFIOWriter writer,
               RIFFIOSeeker seeker,
               RIFFIOTeller teller)
/*
 * RIFFFIOFileInit() should be called immediately after NIFFIOFileNew()
 *
 * ENTRY
 * -----
 * 
 * <prf> :
 *   is the RIFFIOFile that is being initialized.
 *
 * <pvUserFile> :
 *   points to a user representation of a file that will be passed to
 *   user-defined reader, writer, seeker, and teller functions.
 * 
 * <reader>,<writer>,<seeker>,<teller> :
 *   are user-defined functions that perform operations on <*pvUserFile>.
 */
/***************************************************************************/
{

    /* Check arguments */
    assert(prf != 0);

    /* Assign file and I/O methods */
    prf->pvUserFile = pvUserFile;

    prf->read = reader;
    prf->write = writer;
    prf->seek = seeker;
    prf->tell = teller;

    /* 
     * We don't know what the form type is
     * until we read the first (form) chunk.
     * 
     * This should _not_ be done in RIFFIOFileNew() because the 
     * user could call us more than once.
     */
    prf->formType = RIFFIO_FORM_UNKNOWN;
    
    return RIFFIO_OK;
}

/***************************************************************************/
/*
 * RIFFIOFileDelete
 * ================
 * Free the memory allocated to a RIFFIOFile.
 */
void
RIFFIOFileDelete(RIFFIOFile *prf)
/***************************************************************************/
{
    assert (prf != 0);
    
    free(prf);
    
}


/***************************************************************************/
/*
 * RIFFIOFileGetFormType
 * =====================
 * Return the form type (UNKNOWN, RIFF, or RIFX) 
 * of a RIFFIOFile
 */
RIFFIOFormType 
RIFFIOFileGetFormType(RIFFIOFile *prf)
/*
 * The form type is unknown until the first form chunk of a RIFFIOFile
 * is descended into or created.
 */
/***************************************************************************/
{
        assert(prf != 0);
        return prf->formType;
}

/***************************************************************************/
/*
 * RIFFIORead
 * ==========
 * Read bytes from a RIFF file.
 */
long 
RIFFIORead(RIFFIOFile *prf, void *pvBuffer, long nBytes)
/*
 * ENTRY
 * -----
 * T <*pvBuffer> must be allocated to hold <nBytes>.
 *
 * EXIT
 * ----
 * - Up to <nBytes> have been read from <*prf> into <*pvBuffer>. 
 *
 * RETURN
 * ------
 *     *   The number of bytes actually read,
 *     *   0 on end of file,
 *     *  -1 on error.
 */
/***************************************************************************/
{
    long result;

    /*
     * Check input arguments, make sure nothing is Null
     */
    assert(prf != 0);
    assert(prf->read != 0);
    assert(pvBuffer != 0);

    /*
     * Perform the read
     */
    result = prf->read(prf->pvUserFile, pvBuffer, nBytes);

    return result;
}

/***************************************************************************/
/*
 * RIFFIOWrite
 * ===========
 * Write bytes to a RIFF file.
 */
long 
RIFFIOWrite(RIFFIOFile *prf, void *pvBuffer, long nBytes)
/*
 * Write <nBytes> from <*pvBuffer> to <*prf>.
 *
 * RETURN
 * ------
 * The number of bytes actually written.
 */
/***************************************************************************/
{
    long bytesWritten;

    /*
     * Check input arguments, make sure nothing is Null
     */
    assert(prf != 0);
    assert(prf->write != 0);
    assert(pvBuffer != 0);

    /*
     * Perform the write.
     * It's not up to us to check the result.
     */
    bytesWritten = prf->write(prf->pvUserFile, pvBuffer, nBytes);

    return bytesWritten;

}

/***************************************************************************/
/*
 * RIFFIOSeek
 * ==========
 * Seek to a location in a RIFF file.
 */
RIFFIOSuccess
RIFFIOSeek(RIFFIOFile *prf, RIFFIOOffset offset, RIFFIOSeekOrigin origin)
/* 
 * Seek <offset> bytes relative to <origin>.
 *
 * <origin> may be one of
 *
 *     RIFFIO_SEEK_SET:
 *       beginning of file
 *
 *     RIFFIO_SEEK_CUR:
 *       current file position
 *    
 *     RIFFIO_SEEK_END:
 *       end of file
 * 
 * ERRORS
 * ------ 
 * On failure, the file position is undefined.
 */
/***************************************************************************/
{

    /*
     * Check input arguments, make sure nothing is Null
     */
    assert(prf != 0);
    assert(prf->seek != 0);
    assert(origin >= RIFFIO_SEEK_SET);
    assert(origin <= RIFFIO_SEEK_END);

    /*
     * Perform the seek 
     */
    return  prf->seek(prf->pvUserFile, offset, origin);

}

/***************************************************************************/
/*
 * RIFFIOTell
 * ==========
 * Return the current RIFF file position.
 */
long
RIFFIOTell(RIFFIOFile *prf)
/*
 * A file's position is measured in bytes from the
 * beginning of the file.
 */
/***************************************************************************/
{
    long result;

    /*
     * Check input arguments, make sure nothing is Null
     */
    assert(prf != 0);

    /*
     * Perform the tell
     */
    result = prf->tell(prf->pvUserFile);

    return result;

}





